### ISSUE TITLE
Training gives N/A when accuracy is reported

### ISSUE THREAD

**Issue by ChristianEschen** [Kamnitsask/deepmedic#54]
Hi 

I have problems training on my own dataset. Some batches gives NaN's in dice score. 
Have you experienced this before? 
Is it possible to incorporate gradient clipping?
If so which script am I supposed to change.

**Comment by Kamnitsask** [comment]
Hello Christian, I hope you are well.

Are the NaNs in dice score only? And then, are they given after a subepoch as calculated on training/validation samples, or when the model actually goes over the whole validation volumes and segments them?

Because some NaNs are probably intentional. Let me explain.

1) If the NaNs are a result of a real problem during training, such as division by zero or something blowing up to infinity in the network (such as gradients), then you should see the TRAINING cost going to NaN and all the dices to go to either 0 or NaN, for ever (and never recover). If this happens it's indeed a problem. Just check the "cost" reported every subepoch for the TRAINING (not validation). If it's consecutively NaN, then this is the case and we can talk further about what may be going wrong (I doubt clipping is the best solution).
2) If the NaNs appear only for the Dice of a class, or just "sometimes" at the end of a subepoch (especially during "validation" and "not" training), then the reason may simply be that the class is so little represented, that when the validation-samples for this subepoch were extracted randomly, they captured literally no voxel from that class. This happens in many problems with small classes, and it's fine. The N/A in that  case is output actually by me and not theano, to account for that case when I report accuracy.
3) the N/A can also appear during full inference over the validation/test volumes, for the subjects that do not have a class at all.

If you send me your log files at konstantinos.kamnitsas12@imperial.ac.uk I may be able to easier provide some help.

Let me know what's the case out of the two above so I may help further

cheers,
Kostas

**Comment by ChristianEschen** [comment]
Mail has been sent to you.
However, I guess it caused by the fact that my MASK is quite big leading to a lot of samples with some classes pporly represented.

**Comment by Kamnitsask** [comment]
Below I provide a rephrased reply as sent via email, for any other users that may have this issue.

The N/A can be found in the reported DSC metric when the model fully segments the 134 validation subjects (every 5 epochs by default), and often in the sensitivity metric reported on validation samples every subepoch.

This is different from a NaN from theano, and has nothing to do with gradients exploding etc.

This N/A is reported by me in 2 cases.

a) When the DSC is reported after fully segmenting a subject, in the case where in the Ground truth labels for that subject a certain class does not exist. Eg, if a subject does not have at all an abnormality that we try to segment.

b) In the reported sensitivity/precision metric every subepoch, when by chance no randomly extracted validation sample (patch) was extracted from a particular class (eg the abnormalities are so small that when we extract with uniform sampling 5000 patches to check accuracy at that subepoch, no nodule-patch is randomly sampled.)

Both the above do NOT influence training. They are just inconveniences when it comes to reporting metrics. It's quite difficult to design a paradigm that works nicely for any single use-case, I hope you understand.

Kind regards,
Kostas

**Comment by ChristianEschen** [comment]
Dear Konstantinos


Here are the segmentation image from the test set.

As far as I can see the last DICE metric is:

DICE1=[ 0.0432 0.0432 ] DICE2=[ 0.0917 0.0917 ] DICE3=[ 0.0918 0.0918 ]. In the literature I have found that other lung nodule segmentation networks yields approximately 0.8 in validation DSC using FCN networks in 2D.



The performance of deepmedic is not that could right now. The background image is the probability map of not lung nodule and the tumor image is the probability of lung nodule.

It is possible for me to work out a better lung mask, or I could simply not use mask for the training.

What is your advice for better performance?


best regards

Christian Eschen

________________________________
Fra: Konstantinos Kamnitsas <notifications@github.com>
Sendt: 13. april 2017 17:22:37
Til: Kamnitsask/deepmedic
Cc: ChristianEschen; Author
Emne: Re: [Kamnitsask/deepmedic] Training gives NaN's (#54)


Below I provide a rephrased reply as sent via email, for any other users that may have this issue.

The N/A can be found in the reported DSC metric when the model fully segments the 134 validation subjects (every 5 epochs by default), and often in the sensitivity metric reported on validation samples every subepoch.

This is different from a NaN from theano, and has nothing to do with gradients exploding etc.

This N/A is reported by me in 2 cases.

a) When the DSC is reported after fully segmenting a subject, in the case where in the Ground truth labels for that subject a certain class does not exist. Eg, if a subject does not have at all an abnormality that we try to segment.

b) In the reported sensitivity/precision metric every subepoch, when by chance no randomly extracted validation sample (patch) was extracted from a particular class (eg the abnormalities are so small that when we extract with uniform sampling 5000 patches to check accuracy at that subepoch, no nodule-patch is randomly sampled.)

Both the above do NOT influence training. They are just inconveniences when it comes to reporting metrics. It's quite difficult to design a paradigm that works nicely for any single use-case, I hope you understand.

Kind regards,
Kostas

**Comment by Kamnitsask** [comment]
Hi,

Please lets keep the issues here for technical questions / bugs / features regarding the package. I am super-happy to receive feedback and questions about its usage, but in an email.

Please, could you confirm that certain subjects from your database do not have a foreground class at all? (such as LIDC-IDRI-0361, which in your logs gave a N/A in the DSC calculation). This would confirm the reason why N/A appeared and we will close this issue.

Kostas

...
...
...

((( Short note on performance on your task, only because it was brought up. Further discussion/recommendations only in email. Lets keep the rest of the issue for the technical N/A issue, which would help every user, not task-related: 
I am not up to date with nodule segmentation nor worked on it. But as far as I know, pipelines include a candidate detection step (eg segmenter) and a false positive reduction classifier. You could do the 1st with deepmedic but I cannot support you. Methods tend to do much fiddling on this task, eg calibrated training sample distributions, thresholds, candidate merging etc, so you ll need to get your hands dirty. I cannot find anywhere a publication with DSC reported for this 1st step, as methods tend to oversegment a lot. That's why you need the second false positive reduction classifier. Rather, sensitivity, precision, ROC is reported. I doubt a one-step segmenter such as a single FCN can give 80% DSC for this. Please point me to a publication with such a number reported if you have one, I would be interested. Finally, I absolutely not expect, nor hope, nor can I support this software to work for every problem out of the box. Or I would be millionaire :) Now, lets please keep the rest of this issue only for the N/A problem, and rest in emails :-) )))

**Comment by ChristianEschen** [comment]
Alright. 
Lets get back to the original issue.
The issue was related to the fact that some of the images did not contain any ground truth label.
I close the issue.
